internal/internal_worker_base.go

```diff
@@ -332,61 +356,101 @@ func (bw *baseWorker) runPoller() {
 	defer bw.stopWG.Done()
 	bw.metricsHandler.Counter(metrics.PollerStartCounter).Inc(1)
 
+	ctx, cancelfn := context.WithCancel(context.Background())
+	defer cancelfn()
+	reserveChan := make(chan *SlotPermit)
+
 	for {
+		bw.stopWG.Add(1)
+		go func() {
+			defer bw.stopWG.Done()
+			s, err := bw.slotSupplier.ReserveSlot(ctx, &bw.options.slotReservationData)
+			if err != nil {
+				if !errors.Is(err, context.Canceled) {
+					bw.logger.Error(fmt.Sprintf("Error while trying to reserve slot: %v", err))
+					reserveChan <- nil
```

Comment:
This can deadlock if the worker never pulls from `reserveChan`, because it is stopping. you need to select with ctx.Done(). I think it would be better to just retry the `ReserveSlot ` call in this goroutine rather then coordinating with the outside goroutine, but your call

Reply:
I tried it and turned out to not be much simpler. You still need two loops and was giving me some nonobvious problems.